### DISCUSSION TITLE
@bimusiek

### DISCUSSION THREAD

**Discussion by georgehandrew** [pymupdf/PyMuPDF#1471] [Q&A]
@bimusiek 
I have added functionality to solve your problem - hopefully in a fairly elegant way:

* Changed `Page.getImageList()` and `Document.getPageImageList()` to provide the xref of the soft-image mask `/SMask` parameter as the second item of each image entry (it was useless before anyway). If this number is positive, then the image itself should be specially treated for creating PNGs.
* If `[xref, smask,..., "Im2"]` is an entry of `getPageImageList()` and smask is positive, then the following statements should create the original `Im2` png:

```
pix1 = fitz.Pixmap(doc, xref)     # pixmap without alpha channel
pix2 = fitz.Pixmap(doc, smask)    # this contains the alpha values in its samples
pix3 = fitz.Pixmap(pix1)     # copy of pix1 with alpha channel added (new constructor)
pix3.setAlpha(pix2.samples)     # fill its alpha values with pix2 samples (new method)
pix3.writePNG("Im2.png")     # this should look right now!
```

Pixmap pix3 should now reflect the original image - and be what you were missing ... let me know your experiences.

_Originally posted by @JorjMcKie in https://github.com/pymupdf/PyMuPDF/issues/98#issuecomment-333533975_

**Comment by georgehandrew** [comment]
Hi JorjMcKie and all,

[aa.pdf](https://github.com/pymupdf/PyMuPDF/files/7737303/aa.pdf)
This attached scanned pdf file, smask =0, and the extracted png white-black color reversed.
Do you have any solution?
Thanks

**Reply by JorjMcKie** [reply]
From the pixmap you know it actually is a mask:
```python
>>> pix = fitz.Pixmap(doc, 8)  # image xref = 8
>>> pix  # colorspace = None
Pixmap(None, IRect(0, 0, 3296, 2560), 1)
>>> # make a new, empty, gray, non-alpha pixmap with same dimensions
>>> pix1 = fitz.Pixmap(fitz.csGRAY, pix.irect, 0)
>>> pix1
Pixmap(DeviceGray, IRect(0, 0, 3296, 2560), 0)
>>> # fill it with black
>>> pix1.set_rect(pix.irect, [0])
True
>>> # and use the original pix at xref 8 as its mask
>>> pix2 = fitz.Pixmap(pix1, pix)
>>> pix2.save("image.png")
>>> # this will make a decent looking image
```
